src/main/java/ui/components/pickers/LabelPickerDialog.java

```diff
@@ -3,47 +3,171 @@
 import java.util.stream.Collectors;
 
+import com.sun.javafx.tk.FontLoader;
+import com.sun.javafx.tk.Toolkit;
+
+import ui.UI;
+
+/**
+ * A Presenter class to synchronize model and view  after each user action
+ * @author Jin
+ *
+ */
 public class LabelPickerDialog extends Dialog<List<String>> {
 
-    private static final int VBOX_SPACING = 105; // seems like some magic number
+    private static final int VBOX_SPACING = 105;
     private static final int ELEMENT_MAX_WIDTH = 400;
+    private static final String EXCLUSIVE_DELIMITER = ".";
+    private static final String NONEXCLUSIVE_DELIMITER = "-";
 
     private final LabelPickerUILogic uiLogic;
-    private final TextField textField;
-    private final FlowPane topPane;
-    private final VBox bottomBox;
+    private final List<TurboLabel> repoLabels;
+    private final Set<String> repoLabelsString;
+    private final Map<String, Boolean> groups;
+    private final TurboIssue issue;
+    
+    private List<String> initialLabels;
+    private List<String> addedLabels;
+    private List<String> removedLabels;
+    private List<String> matchedLabels;
+    private List<String> finalLabels;
+    
+
+    @FXML
+    private VBox mainLayout;
+    @FXML
+    private Label title;
+    @FXML
+    private FlowPane assignedLabels;
+    @FXML
+    private TextField queryField;
+    @FXML
+    private VBox feedbackLabels;
 
     LabelPickerDialog(TurboIssue issue, List<TurboLabel> repoLabels, Stage stage) {
+        this.repoLabels = repoLabels;
+        this.repoLabelsString = repoLabels.stream()
+                .map(TurboLabel::getActualName)
+                .collect(Collectors.toSet());
+        this.issue = issue;
+        LabelPickerState state = new LabelPickerState(new HashSet<String>(issue.getLabels()));
+        uiLogic = new LabelPickerUILogic();
+
+        // Logic initialisation
+        this.groups = initGroups(repoLabels);
+
         // UI creation
+        initUI(stage, issue);
+        updateUI(state);
+        setupEvents(stage);
+        Platform.runLater(queryField::requestFocus);
+    }
+
+    // Init Assets 
+    private Map<String, Boolean> initGroups(List<TurboLabel> repoLabels) {
+        Map<String, Boolean> groups = new HashMap<>();
+        repoLabels.forEach(label -> {
+            if (label.getGroup().isPresent() && !groups.containsKey(label.getGroup().get())) {
+                groups.put(label.getGroup().get(), label.isExclusive());
+            }        
+        });
+        return groups;
+    }
+
+    // Initilisation of UI
+
+    private void initialiseDialog(Stage stage, TurboIssue issue) {
+        initOwner(stage);
+        initModality(Modality.APPLICATION_MODAL); // TODO change to NONE for multiple dialogs
+        setTitle("Edit Labels for " + (issue.isPullRequest() ? "PR #" : "Issue #") +
+                issue.getId() + " in " + issue.getRepoId());
+    }
+
+    private void positionDialog(Stage stage) {
+        if (!Double.isNaN(getHeight())) {
```

Comment:
When can this field be NaN? If it's an error condition, it would be good to log it, or add a comment explaining why it's ignored. If it's expected, maybe somewhere higher up you could use an Optional?